command/sub8_missions/missions/collect_classify.py

```diff
@@ -0,0 +1,387 @@
+        # Rotate 180 if we're not facing pinger
+        m = trns.quaternion_matrix(self.sub.move._pose.orientation).dot(np.array([0.1, 0, 0]))
+        if np.linalg.norm(m - self.pinger_position) > np.linalg.norm(move._pose.position - self.pinger_position):
+            move = move.yaw_left(np.pi)
+
+        move = move.backward(1.5)
+        yield move.go()
+
+        # Open arm
+        yield self.sub.actuators.gripper_open()
+
+        # Move forward
+        move = move.forward(1.0)
+        yield move.go()
+        yield self.sub.acutators.gripper_close()  # Close arm around pipe
+
+        yield self.sub.move.up(2.0).go()  # move up to lift pipe
+
+    @util.cancellableInlineCallbacks
+    def get_oval(self):
+        '''
+        Asks oval perception for the pose of each unfound oval. Returns the Oval
+        object of the first one found, or None if none are found.
+        '''
+        for o in self.ovals:
+            res = self.sub.vision_proxies.table_ovals.get_pose(target=o.color)
+            if res.found:
+                # TODO: check frame, convert if needed
+                o.pose = self.vision_response_to_pose(res)
+                self.ovals_found.append(o)
+                self.ovals.remove(o)
+                defer.returnValue(o)
+        defer.returnValue(None)
+
+    @util.cancellableInlineCallbacks
+    def get_pipe(self):
+        '''
+        Asks pipe perception for the pose of each unfound pipe. Returns the Pipe
+        object of the first one found, or None if none are found.
+        '''
+        for p in self.pipes:
+            res = self.sub.vision_proxies.tower_pipes.get_pose(target=p.color)
+            if res.found:
+                # TODO: check frame, convert if needed
+                p.pose = self.vision_response_to_pose(res)
+                self.pipes_found.append(p)
+                self.ovals.remove(p)
+                defer.returnValue(p)
+        defer.returnValue(None)
+
+    @util.cancellableInlineCallbacks
+    def find_first_oval(self):
+        '''
+        Circle at octogon radius around pinger to find the oval table.
+        If an oval has already been found or if one is found during search pattern,
+        return True. Otherwise, return False.
+        '''
+        if len(self.ovals_found) > 0:
+            defer.returnValue(True)
+
+        origin = self.pinger_position
+        origin[2] = CollectClassifyMission.OVAL_SEARCH_DEPTH
+        circle_done = False
+
+        @util.cancellableInlineCallbacks
+        def circle():
+            yield self.circle(origin, CollectClassifyMission.OCTOGON_RADIUS)
+            circle_done = True  # noqa
+
+        search = circle()
+        while not circle_done:
+            res = yield self.get_oval()
+            if res is not None:
+                search.cancel()
+                defer.returnValue(True)
+        defer.returnValue(False)
+
+    @util.cancellableInlineCallbacks
+    def find_oval(self, color):
```

Comment:
This would probably lead to better results, I will work on this once the oval perception is complete.